userspace/engine/rule_loader.cpp

```diff
@@ -0,0 +1,1098 @@
+	}
+	else if (doc != docs.end())
+	{
+		// end of item is beginning of next doc
+		to = doc->Mark().pos - 4;
+	}
+	else
+	{
+		// end of item is end of file contents
+		to = content.length();
+	}
+	string obj = content.substr(from, to - from);
+	return trim(obj);
+}
+
+template <typename T>
+static bool yaml_is_type(const YAML::Node& v)
+{
+	T t;
+	return v.IsDefined() && v.IsScalar() && YAML::convert<T>::decode(v, t);
+}
+
+template <typename T>
+static void yaml_decode_seq(const YAML::Node& item, vector<T>& out)
+{
+	T value;
+	for(const YAML::Node& v : item)
+	{
+		THROW(!v.IsScalar() || !YAML::convert<T>::decode(v, value),
+			"Can't decode YAML sequence value");
+		out.push_back(value);
+	}
+}
+
+template <typename T>
+static void yaml_decode_seq(const YAML::Node& item, set<T>& out)
+{
+	T value;
+	for(const YAML::Node& v : item)
+	{
+		THROW(!v.IsScalar() || !YAML::convert<T>::decode(v, value),
+			"Can't decode YAML sequence value");
+		out.insert(value);
+	}
+}
+
+// todo(jasondellaluce): this breaks string escaping in lists
+static bool resolve_list(string& cnd, YAML::Node& list)
+{
+	static string blanks = " \t\n\r";
+	static string delims = blanks + "(),=";
+	string new_cnd;
+	size_t start, end;
+	bool used = false;
+	start = cnd.find(list["list"].as<string>());
+	while (start != string::npos)
+	{
+		// the characters surrounding the name must
+		// be delims of beginning/end of string
+		end = start + list["list"].as<string>().length();
+		if ((start == 0 || delims.find(cnd[start - 1]) != string::npos)
+			&& (end >= cnd.length() || delims.find(cnd[end]) != string::npos))
+		{
+			// shift pointers to consume all whitespaces
+			while (start > 0
+				&& blanks.find(cnd[start - 1]) != string::npos)
+			{
+				start--;
+			}
+			while (end < cnd.length()
+				&& blanks.find(cnd[end]) != string::npos)
+			{
+				end++;
+			}
+			// create substitution string by concatenating all values
+			string sub = "";
+			for (auto v : list["items"])
+			{
+				if (!sub.empty())
+				{
+					sub += ", ";
+				}
+				sub += v.as<string>();
+			}
+			// if substituted list is empty, we need to
+			// remove a comma from the left or the right
+			if (sub.empty())
+			{
+				if (start > 0 && cnd[start - 1] == ',')
+				{
+					start--;
+				}
+				else if (end < cnd.length() && cnd[end] == ',')
+				{
+					end++;
+				}
+			}
+			// compose new string with substitution
+			new_cnd = "";
+			if (start > 0)
+			{
+				new_cnd += cnd.substr(0, start) + " ";
+			}
+			new_cnd += sub + " ";
+			if (end <= cnd.length())
+			{
+				new_cnd += cnd.substr(end);
+			}
+			cnd = new_cnd;
+			start += sub.length() + 1;
+			used = true;
+		}
+		start = cnd.find(list["list"].as<string>(), start + 1);
+	}
+	return used;
+}
+
+static void resolve_macros(
+	indexed_vector<pair<YAML::Node,shared_ptr<ast::expr>>>& macros,
+	map<string, bool>& used_macros,
+	shared_ptr<ast::expr>& ast,
+	uint32_t index_visibility,
+	string on_unknown_err_prefix)
```

Comment:
This is not a reference just so you can construct error strings on the fly and pass them in a single step, right?

Reply:
Yeah, exactly